### ISSUE TITLE
Parent directory not in source state when importing zplug

### ISSUE THREAD

**Issue by Quintasan** [twpayne/chezmoi#1205]
## What exactly are you trying to do?

I'm trying to import zplug tarball into chezmoi

## What have you tried so far?

```
quintasan@demonbane:~$ chezmoi import --strip-components 1 --destination ${HOME}/.config/zplug zplug.tar.gz
chezmoi: /home/quintasan/.config/zplug/.devbots: parent directory not in source state
```

## Where else have you checked for solutions?

* [X] I have read [chezmoi's how-to guide](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md), and not found the answer.
* [X] I have read [chezmoi's FAQ](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md), and not found the answer.
* [X] I have searched [chezmoi's reference guide](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md), and not found the answer.
* [X] Other, please give details – Googled for the exact error message

I also tried removing entirety of zplug and starting from scratch.

## Output of any commands you've tried with `--verbose`

```
$ chezmoi import --verbose --strip-components 1 --destination ${HOME}/.config/zplug zplug.tar.gz
chezmoi: /home/quintasan/.config/zplug/.devbots: parent directory not in source state
```

## Output of `chezmoi doctor`

```
$ chezmoi doctor
chezmoi doctor
RESULT   CHECK                MESSAGE
ok       version              v2.0.12, commit 2256a83c51b97f4a326feb66fd3ebdc3f1833a15, built at 2021-05-10T22:46:04Z, built by goreleaser
ok       os-arch              linux/amd64
ok       config-file          /home/quintasan/.config/chezmoi/chezmoi.toml is a file
ok       source-dir           /home/quintasan/.local/share/chezmoi is a directory
ok       suspicious-entries   no suspicious entries found in /home/quintasan/.local/share/chezmoi
ok       dest-dir             /home/quintasan is a directory
ok       shell                found /bin/zsh
ok       editor               found /usr/bin/vi
ok       git-cli              found /home/quintasan/.nix-profile/bin/git, version 2.30.0
ok       merge-cli            found /usr/bin/vimdiff
ok       age-cli              age not found in $PATH
ok       gnupg-cli            found /usr/bin/gpg, version 2.2.19
info     1password-cli        op not found in $PATH
info     bitwarden-cli        bw not found in $PATH
info     gopass-cli           gopass not found in $PATH
info     keepassxc-cli        keepassxc-cli not found in $PATH
info     keepassxc-db         not set
info     lastpass-cli         lpass not found in $PATH
ok       pass-cli             found /usr/bin/pass, version 1.7.3
info     vault-cli            vault not found in $PATH
info     secret-cli           not set
```

**Comment by twpayne** [comment]
chezmoi's documentation can be improved here.

You need to create the directory for zplug in the source state before running chezmoi import, i.e.

    mkdir ~/.local/share/chezmoi/dot_config/zplug

**Comment by twpayne** [comment]
I think that the right thing to do here is for chezmoi to create the parent directories if needed.

**Comment by dhruvinsh** [comment]
I am having a same issue as well for external plugin setup. 

```
[".tmux/plugins/tpm"]
  type = "archive"
  url = "https://github.com/tmux-plugins/tpm/archive/master.tar.gz"
  exact = true
  stripComponents = 1
```

but that `.tmux/plugins` directory doesn't exists hence need to create parent directory manually.

**Comment by twpayne** [comment]
In the short term you can create it manually with:

```console
$ chezmoi cd
$ mkdir -p dot_tmux/plugins/tpm
$ touch dot_tmux/plugins/tpm.keep
```

**Comment by twpayne** [comment]
> I think that the right thing to do here is for chezmoi to create the parent directories if needed.

For info, the reason I didn't implement the creation of the parent directories so far is that chezmoi does not know what attributes (e.g. `private_` or `exact_`) the parent directories should have, hence the requirement for the user to create them.

**Comment by dhruvinsh** [comment]
> In the short term you can create it manually with:
> ```
>$ chezmoi cd
>$ mkdir -p dot_tmux/plugins/tpm
>$ touch dot_tmux/plugins/tpm.keep
>```

did you mean `touch dot_tmux/plugins/tpm/.keep` (see the slash before `.keep`) ?

**Comment by twpayne** [comment]
Thanks for the info. I think this is a bug in chezmoi.

Note to self: multiple equivalent source state entries for the same target state entry should be merged rather than being considered duplicates. [This check](https://github.com/twpayne/chezmoi/blob/master/internal/chezmoi/sourcestate.go#L833) should be replaced with a with a check that only errors out if the target state entries are inconsistent.

**Comment by dhruvinsh** [comment]
similar to this another issue found,
file: .chezmoiexternal.toml
``` toml
[".oh-my-zsh"]
  type = "archive"
  url = "https://github.com/ohmyzsh/ohmyzsh/archive/master.tar.gz"
  exact = true
  stripComponents = 1
[".oh-my-zsh/custom/plugins/zsh-syntax-highlighting"]
    type = "archive"
    url = "https://github.com/zsh-users/zsh-syntax-highlighting/archive/master.tar.gz"
    exact = true
    stripComponents = 1
```
Everything is working as expect with `chezmoi apply`. But now while my config is managed by chezmoi I decided to add aliases in  `.oh-my-zsh/custom/aliases.zsh`. and I wanted to track it by chezmoi but when i try to add it by chezmoi I receive below error

```
> ~ chezmoi add .oh-my-zsh/custom/aliases.zsh
chezmoi: stat /home/ds/.local/share/chezmoi/https:/github.com/ohmyzsh/ohmyzsh/archive/master.tar.gz: no such file or directory
```

**Comment by twpayne** [comment]
Thank you for reporting this.

Right now, chezmoi doesn't support mixing files in the source state with externals, and you're right in reporting it here as part of this issue as I think the underlying fix is the same:

> Note to self: multiple equivalent source state entries for the same target state entry should be merged rather than being considered duplicates. [This check](https://github.com/twpayne/chezmoi/blob/master/internal/chezmoi/sourcestate.go#L833) should be replaced with a with a check that only errors out if the target state entries are inconsistent.

The embedding of the external URL into the path here:

> chezmoi: stat /home/ds/.local/share/chezmoi/https:/github.com/ohmyzsh/ohmyzsh/archive/master.tar.gz: no such file or directory

This is equally a very clear bug and a direct consequence of [a hack I did for externals](https://github.com/twpayne/chezmoi/blob/master/internal/chezmoi/sourcestate.go#L1485) (i.e converting a URL to a relative path).

**Comment by BasixKOR** [comment]
You can work around it by creating nesscary `.keep` files to to force chezmoi to create folders. I've done this in my dotfiles repo and you can see it [here](https://github.com/BasixKOR/.dotfiles/commit/0f4f66d8fea4fffd22f8db4ff5ff08ce528bc954).

**Comment by twpayne** [comment]
The creation of parent directories if needed should be fixed by #1571.

The merging of duplicate equivalent states as described in [this comment](https://github.com/twpayne/chezmoi/issues/1205#issuecomment-912090344) is not yet fixed.

> And oops it seems like this didn't work.

@BasixKOR could you give more details on what exactly didn't work?

**Comment by BasixKOR** [comment]
Yep I was talking about that duplicate equivalent states thing. It worked on my case though.
